File: packages/node-server/src/cache.ts
import { CacheEntry } from './types';

/**
 * In-memory cache for flag values with TTL
 */
export class FlagCache {
  private cache: Map<string, CacheEntry>;
  private ttl: number;

  constructor(ttl: number = 60000) {
    this.cache = new Map();
    this.ttl = ttl;
  }

  /**
   * Get a cached flag value
   */
  get(key: string): boolean | null {
    const entry = this.cache.get(key);
    if (!entry) return null;

    // Check if expired
    if (Date.now() > entry.expiresAt) {
      this.cache.delete(key);
      return null;
    }

    return entry.value;
  }

  /**
   * Set a flag value in cache
   */
  set(key: string, value: boolean, flagId?: string): void {
    this.cache.set(key, {
      value,
      expiresAt: Date.now() + this.ttl,
      flagId,
    });
  }

  /**
   * Invalidate a specific flag or all flags
   */
  invalidate(key?: string): void {
    if (key) {
      this.cache.delete(key);
    } else {
      this.cache.clear();
    }
  }

  /**
   * Clear all cached values
   */
  clear(): void {
    this.cache.clear();
  }
}

File: packages/go-server/cache.go
package savvagent

import (
	"sync"
	"time"
)

// cache provides thread-safe in-memory caching for flag values
type cache struct {
	mu      sync.RWMutex
	entries map[string]cacheEntry
	ttl     time.Duration
}

// newCache creates a new cache with the specified TTL
func newCache(ttl time.Duration) *cache {
	return &cache{
		entries: make(map[string]cacheEntry),
		ttl:     ttl,
	}
}

// Get retrieves a cached value if it exists and hasn't expired
func (c *cache) Get(key string) (bool, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	entry, exists := c.entries[key]
	if !exists {
		return false, false
	}

	// Check if expired
	if time.Now().After(entry.ExpiresAt) {
		return false, false
	}

	return entry.Value, true
}

// Set stores a value in the cache with TTL
func (c *cache) Set(key string, value bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.entries[key] = cacheEntry{
		Value:     value,
		ExpiresAt: time.Now().Add(c.ttl),
	}
}

// Invalidate removes a specific key from the cache
// If key is empty, it clears all entries
func (c *cache) Invalidate(key string) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if key == "" {
		c.entries = make(map[string]cacheEntry)
	} else {
		delete(c.entries, key)
	}
}

// Clear removes all entries from the cache
func (c *cache) Clear() {
	c.Invalidate("")
}
